applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.h

```diff
@@ -0,0 +1,524 @@
+        const unsigned int max_number_of_neighbors = mMapperSettings["max_nodes_in_filter_radius"].GetInt();
+
+        // variable size, fixed capacity vecs
+        std::vector<bool> transform;
+        NodeVector total_neighbor_nodes;
+        std::vector<double> total_list_of_weights;
+        std::vector<double> list_of_weights;
+        transform.reserve(max_number_of_neighbors);
+        total_neighbor_nodes.reserve( max_number_of_neighbors );
+        total_list_of_weights.reserve( max_number_of_neighbors );
+        list_of_weights.reserve( max_number_of_neighbors );
+
+        // fixed size vecs
+        std::vector<double> resulting_squared_distances( max_number_of_neighbors );
+        NodeVector neighbor_nodes( max_number_of_neighbors );
+
+        for(auto& node_i : mrDestinationModelPart.Nodes())
+        {
+            transform.clear();
+            total_neighbor_nodes.clear();
+            total_list_of_weights.clear();
+
+            auto search_nodes = mpSymmetry->GetDestinationSearchNodes(node_i.GetValue(MAPPING_ID));
+            unsigned int total_number_of_neighbors = 0;
+            double total_sum_of_weights = 0;
+            for (auto& pair_i : search_nodes) {
+                NodeType search_node(0, pair_i.first);
+                const unsigned int number_of_neighbors = mpSearchTree->SearchInRadius(search_node,
+                                                                                filter_radius,
+                                                                                neighbor_nodes.begin(),
+                                                                                resulting_squared_distances.begin(),
+                                                                                max_number_of_neighbors );
+                total_number_of_neighbors += number_of_neighbors;
+                transform.resize(total_number_of_neighbors, pair_i.second);
+
+                list_of_weights.clear();
+                list_of_weights.resize( number_of_neighbors, 0.0 );
+                ComputeWeightForAllNeighbors( search_node, neighbor_nodes, number_of_neighbors, list_of_weights, total_sum_of_weights );
+
+                total_list_of_weights.insert(total_list_of_weights.end(), list_of_weights.begin(), list_of_weights.begin()+number_of_neighbors);
+                total_neighbor_nodes.insert(total_neighbor_nodes.end(), neighbor_nodes.begin(), neighbor_nodes.begin()+number_of_neighbors);
+            }
+
+            if(total_number_of_neighbors >= max_number_of_neighbors)
+                KRATOS_WARNING("ShapeOpt::MapperVertexMorphingSymmetric") << "For node " << node_i.Id() << " and specified filter radius, maximum number of neighbor nodes (=" << max_number_of_neighbors << " nodes) reached!" << std::endl;
+
+            FillMappingMatrixWithWeights( node_i, total_neighbor_nodes, total_number_of_neighbors, total_list_of_weights, transform, total_sum_of_weights );
+        }
+    }
```

Comment:
This section also can be OMP parallelized with critical sections ryt?

Reply:
it could, but then the assembly of the matrix does not happen in row by row anymore, that requires a lot of reallocations, copying as far as i understood. But i am not sure tbh.

Reply:
mm.... then we can keep it as it is :)

Reply:
I think it  can definitely be done, but requires care

Bascially you need to allocate the positions first to avoid constatn reallocations
after all this is how the B&S does it

I am currently working on moving the MappingApp to the Core, maybe you could use some functionalities of it in the future

Reply:
I tried it and its actually quite a bit faster.